### DISCUSSION TITLE
Adding Roles to Next-Auth with Firebase Email & Password using CredentialsProvider.

### DISCUSSION THREAD

**Discussion by Adhishtanaka** [nextauthjs/next-auth#10209] [Help]
Hey guys,

Hope you're all having a great day! I'm currently working on setting up authentication for my project using Next-Auth and Firebase's email/password setup. I'm trying to add role-based authentication to control access to different parts of my app.

I've tried adding a profile() callback to the providers' configuration to determine user roles. However, I hit a snag. It's throwing an error that says: "Object literal may only specify known properties, and 'profile' does not exist in type 'UserCredentialsConfig<{}>'".

I've been scratching my head over this for a while now and couldn't find a straightforward solution in the docs or online.

Could someone lend me a hand and guide me through this? Any pointers or examples on how to add role-based authentication with Next-Auth and Firebase email/password would be awesome!

Thanks a bunch for your help!
  
```
//app\api\auth\[...nextauth]\route.ts 
//i'm using next.js 14 with turbo and approuter

import NextAuth from "next-auth"
import CredentialsProvider from "next-auth/providers/credentials";
import {signInWithEmailAndPassword} from 'firebase/auth';
import auth from '@/app/lib/auth';


export const authOptions = {
 
  secret: process.env.AUTH_SECRET,
  pages: {
    signIn: '/signin'
  },
  session: {
    strategy: "jwt" as const,
    maxAge: 3600,
  },

  providers: [
    CredentialsProvider({
      name: 'Credentials',
     // profile(profile) {
     //   return { role: profile.role ?? "user", ... }
     //  },
      credentials: {},
      async authorize(credentials): Promise<any> {
        return await signInWithEmailAndPassword(auth, (credentials as any).email || '', (credentials as any).password || '')
          .then(userCredential => {
            if (userCredential.user) {
              console.log(userCredential.user)
              return userCredential.user;
            }
            return null;
          })
          .catch(error => (console.log(error)))
  .catch((error) => {
    console.log(error);
  });
      }
    })
  ],
  callbacks: {
    jwt: async ({ token, user }: { token: any, user: any }) => {
      if (user) token.role = user.role;
      return token;
    },
    session: async ({ session, token }: { session: any, token: any }) => {
      session.user.role = token.role;
      return session;
    }
  }
}  
const handler = NextAuth(authOptions)
export { handler as GET, handler as POST}
```

**Comment by mvn-h24** [comment]
Hello, @Adhishthana. I think you don't need to use the `profile()` callback or anything like that to authorize the user's access for role-based actions or content of app.

Next-auth performs work on user authentication, automates and centralizes work with the user's session.

As myself i have not encountered a task like yours, but as far as I understand, a library like [cals.js](https://casl.js.org/) - that what you really need. Use it after the user's session is ready for role-based analysis inside the application.

**Reply by Adhishtanaka** [reply]
Thank you @mvn-h24 for your suggestion. I'll definitely look into cals.js as you recommended and assess whether it fits the requirements of the task at hand. Your insight has been helpful in considering alternative approaches & Once again, thank you for taking the time to share your thoughts and provide guidance.
